File: backend/controllers/contactController.js
import { Contact } from '../models/Contact.js';

// Temporary in-memory log for when DB is not connected
const memoryContacts = [];

export const createContact = async (req, res) => {
  try {
    const { name, email, subject, message } = req.body;
    if (!name || !email || !message) {
      return res.status(400).json({ success: false, error: 'Name, email, and message are required.' });
    }

    try {
      const newContact = new Contact({ name, email, subject: subject || 'General Inquiry', message });
      await newContact.save();
      return res.status(201).json({
        success: true,
        message: 'Message saved to MongoDB successfully!',
        data: newContact
      });
    } catch (dbErr) {
      const fallbackEntry = { id: Date.now(), name, email, subject, message, createdAt: new Date() };
      memoryContacts.push(fallbackEntry);
      return res.status(201).json({
        success: true,
        message: 'Message received and logged in server memory!',
        data: fallbackEntry
      });
    }
  } catch (error) {
    return res.status(500).json({ success: false, error: 'Server error processing contact message.' });
  }
};

export const getContacts = async (req, res) => {
  try {
    const contacts = await Contact.find().sort({ createdAt: -1 });
    return res.json({ success: true, count: contacts.length, data: contacts });
  } catch (error) {
    return res.json({ success: true, count: memoryContacts.length, data: memoryContacts });
  }
};

File: backend/src/main/java/com/jyoti/portfolio/controller/ContactController.java
package com.jyoti.portfolio.controller;

import com.jyoti.portfolio.dto.ContactDTO;
import com.jyoti.portfolio.model.Contact;
import com.jyoti.portfolio.service.ContactService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/contact")
public class ContactController {

    private final ContactService contactService;

    @Autowired
    public ContactController(ContactService contactService) {
        this.contactService = contactService;
    }

    @PostMapping
    public ResponseEntity<Map<String, Object>> createContact(@RequestBody ContactDTO contactDTO) {
        Map<String, Object> response = new HashMap<>();

        if (contactDTO.getName() == null || contactDTO.getName().trim().isEmpty() ||
            contactDTO.getEmail() == null || contactDTO.getEmail().trim().isEmpty() ||
            contactDTO.getMessage() == null || contactDTO.getMessage().trim().isEmpty()) {
            response.put("success", false);
            response.put("error", "Name, email, and message are required.");
            return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(response);
        }

        Contact contact = new Contact(
            contactDTO.getName(),
            contactDTO.getEmail(),
            contactDTO.getSubject(),
            contactDTO.getMessage()
        );

        try {
            Contact savedContact = contactService.saveContact(contact);
            response.put("success", true);
            response.put("message", "Message saved to database successfully!");
            response.put("data", savedContact);
            return ResponseEntity.status(HttpStatus.CREATED).body(response);
        } catch (Exception dbErr) {
            // Fallback in-memory entry created during saveContact throw
            response.put("success", true);
            response.put("message", "Message received and logged in server memory!");
            response.put("data", contact);
            return ResponseEntity.status(HttpStatus.CREATED).body(response);
        }
    }

    @GetMapping
    public ResponseEntity<Map<String, Object>> getContacts() {
        List<Contact> contacts = contactService.getAllContacts();
        Map<String, Object> response = new HashMap<>();
        response.put("success", true);
        response.put("count", contacts.size());
        response.put("data", contacts);
        return ResponseEntity.ok(response);
    }
}
